### ISSUE TITLE
r10k hangs with 100% CPU usage

### ISSUE THREAD

**Issue by fraenki** [puppetlabs/r10k#768]
When running r10k it sometimes gets stuck with 100% CPU usage. This happens on two separate hosts. It's not reproducable, but happens *at least* once per day.

Environment:
```
$ /opt/puppetlabs/puppet/bin/r10k version
r10k 2.5.5

# host 1: Debian 8.9
Linux 4.9.0-0.bpo.3-amd64 #1 SMP Debian 4.9.30-2+deb9u2~bpo8+1 (2017-06-27) x86_64 GNU/Linux

# host 2: Ubuntu 16.04
Linux 4.4.0-93-generic #116-Ubuntu SMP Fri Aug 11 21:17:51 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
```

Command:
```
/opt/puppetlabs/puppet/bin/r10k deploy environment -vp
```

Output of pstree:
```  │           └─r10k,32154 /opt/puppetlabs/puppet/bin/r10k deploy environment -vp
  │               ├─r10k,1721 /opt/puppetlabs/puppet/bin/r10k deploy environment -vp
  │               ├─{pump.rb:30},1719
  │               ├─{pump.rb:30},1720
  │               └─{ruby-timer-thr},1722
```

Output of strace:
```
# strace -p 1721
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0
sched_yield()                           = 0

# strace -p 1720
Process 1720 attached
select(10, [9], [], [], {0, 782361})    = 0 (Timeout)
fcntl(9, F_GETFL)                       = 0x800 (flags O_RDONLY|O_NONBLOCK)
read(9, 0x7f7594afc950, 4096)           = -1 EAGAIN (Resource temporarily unavailable)
select(10, [9], [], [], {1, 600000})    = 0 (Timeout)
fcntl(9, F_GETFL)                       = 0x800 (flags O_RDONLY|O_NONBLOCK)
read(9, 0x7f7594afda80, 4096)           = -1 EAGAIN (Resource temporarily unavailable)
select(10, [9], [], [], {1, 600000})    = 0 (Timeout)
fcntl(9, F_GETFL)                       = 0x800 (flags O_RDONLY|O_NONBLOCK)
read(9, 0x7f7594afebb0, 4096)           = -1 EAGAIN (Resource temporarily unavailable)
select(10, [9], [], [], {1, 600000})    = 0 (Timeout)
```

**Comment by SamK** [comment]
I'm experiencing the same issue and commented about it on bug #650.

**Comment by strickra** [comment]
As am I / as have I.

**Comment by cova-fe** [comment]
Same issue here; I can add that it can be seen easily when dealing with many repos; if r10k manages only one it is very difficult to reproduce the issue.

**Comment by tutelacooldouce** [comment]
hello , I have exactly the same behavior,.  any update ?

**Comment by cova-fe** [comment]
I can add this bit: In an environment that is quite prone to this issue, I see also this happening:
(ruby provided by puppet5)
```
 [ASYNC BUG] (6)
 EPIPE
 ruby 2.4.2p198 (2017-09-14 revision 59899) [x86_64-linux]
 [NOTE]
 You may have encountered a bug in the Ruby interpreter or extension libraries.
 Bug reports are welcome.
 For details: http://www.ruby-lang.org/bugreport.html
 
 sh: line 1: 32469 Aborted                 (core dumped) r10k deploy environment -pv
```

**Comment by Scharfenberg** [comment]
I'm having the same issue (in 30% of all r10k runs):
```
[ASYNC BUG] (6)
EPIPE

ruby 2.4.4p296 (2018-03-28 revision 63013) [x86_64-linux]

[NOTE]
You may have encountered a bug in the Ruby interpreter or extension libraries.
Bug reports are welcome.
For details: http://www.ruby-lang.org/bugreport.html

Aborted
```
r10k version: 2.6.2
git version: 2.11.0
on Debian stretch

**Comment by MikaelSmith** [comment]
Anyone want to give https://rbspy.github.io a go on a hung invocation, or try digging in with some debugging statements? From a quick look through reports, I don't see an obvious starting point to reproduce this issue.

**Comment by Magisus** [comment]
FWIW we also saw this quite a bit in CI, mostly on Debian 9, though very occasionally on other platforms. We merged something that seemed to help quite a bit in our pipelines, see https://github.com/puppetlabs/r10k/pull/866. This was released in r10k 2.6.5/3.0.3. If you can, maybe try upgrading and see if the issue is fixed.

Specifically, we saw the `ASYNC BUG`, and fixed it with the above. I don't think hanging or high CPU usage were specifically noticed, mostly because it often occurred during automated testing.

**Comment by fraenki** [comment]
I haven't experienced this issue in a while, so closing this issue now.
